File: 프로그래머스/2/42577. 전화번호 목록/전화번호 목록.cpp
#include <string>
#include <iostream>
#include <vector>
#include <unordered_map>

using namespace std;

unordered_map<string,int> m;

bool solution(vector<string> phone_book) {
    
    for(int i=0;i<phone_book.size();i++) {
        m[phone_book[i]]=i+1;
    }
    
    for(int i=0;i<phone_book.size();i++) {
        string temp="";
        for(char temp_Char:phone_book[i]) {
            temp+=temp_Char;
        
            if(m[temp]!=i+1 and m[temp]!=0) {
                
                return false;
            }
        }
    }
    return true;;
    
}

File: 프로그래머스/2/42577. 전화번호 목록/전화번호 목록.java
import java.util.*;

class Solution {
    
    static HashMap<String,Integer> map = new HashMap<>();
    
    public boolean solution(String[] phone_book) {
        boolean answer = true;
        
        for(int i=0;i<phone_book.length;i++) {
            map.put(phone_book[i],i+1);
        }
        
        for(int i=0;i<phone_book.length;i++) {
            String temp="";
            
            for(int j=0;j<phone_book[i].length();j++) {
                temp+=phone_book[i].charAt(j);
                
                if(map.get(temp)==null) continue;
                if(map.get(temp)!=i+1) answer=false;
            }
        }
        return answer;
    }
}